### ISSUE TITLE
Error when using v4

### ISSUE THREAD

**Issue by noinkling** [gaearon/react-hot-loader#728]
### Description

Error on hot reload when rendering certain parts of the app.

### Actual behavior

~~Blank page, error in browser console:~~ (fixed in beta 4)

> ```
> hotReplacementRender.js:118 Uncaught TypeError: Cannot read property 'length' of undefined
>     at mergeInject (hotReplacementRender.js:118)
>     at flow.forEach (hotReplacementRender.js:162)
>     at Array.forEach (<anonymous>)
>     at hotReplacementRender (hotReplacementRender.js:139)
>     at next (hotReplacementRender.js:151)
>     at flow.forEach (hotReplacementRender.js:168)
>     at Array.forEach (<anonymous>)
>     at hotReplacementRender (hotReplacementRender.js:139)
>     at next (hotReplacementRender.js:151)
>     at flow.forEach (hotReplacementRender.js:168)
> ```

> ```
> react-dom.development.js:9747 The above error occurred in the <AppContainer> component:
>     in AppContainer (created by HotExportedApp)
>     in HotExportedApp
>
> Consider adding an error boundary to your tree to customize error handling behavior.
> Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
> ```

Depending on the file I modify I also sometimes get this error (before the others), but I'm not sure if it's related:

> ```
> React-stand-in: Updated class  Connect(SomeContainer) contains native or bound function  
> notifyNestedSubs ƒ notifyNestedSubs() {
>     this.listeners.notify();
>   } . Unable to reproduce, use arrow functions instead.
> ```

### Environment

React Hot Loader version: 4.0.0-beta.3

React version: 16.2.0

`node -v`: 9.3.0

`yarn -v`: 1.3.2

Operating system: Windows 10

Browser and version: Chrome 63

Webpack is running via webpack-dev-middleware in WSL (no errors in terminal).

**Comment by theKashey** [comment]
Hooray! The first issue!
Will fix it today.

**Comment by theKashey** [comment]
`React-stand-in: Updated class  Connect(SomeContainer) contains native or bound function  
notifyNestedSubs ƒ notifyNestedSubs() {
    this.listeners.notify();
  } . Unable to reproduce, use arrow functions instead.`
Any ideas why it can be `bound`?
Any code example to reproduce?

**Comment by noinkling** [comment]
I have no idea. `notifyNestedSubs` isn't defined by me, it seems to be a react-redux internal thing. I'm also not currently using `.bind` anywhere in my code.

I will try and reproduce when I get the chance, but it might be tough to extract from the app.

**Comment by theKashey** [comment]
@noinkling - could you try to debug this issue?
Just place a breakpoint at the to the place this error if from and report why that code branch is executed.
It just compares function1.length vs function2.length (both must be zero for this case), and checking that this function does exist in prototype (it should). And first of all - it does check that function is not `bound` aka _[native code]_.
It also should not be bound - something else cast a side effect on it.

**Comment by fwh1990** [comment]
I got the same error message:
```js

React-stand-in: Updated class  Connect(SomeContainer) contains native or bound function  
notifyNestedSubs ƒ notifyNestedSubs() {
    this.listeners.notify();
  } . Unable to reproduce, use arrow f

```

But rhl still worked.

**Comment by theKashey** [comment]
Ok, I've found a problem. RHL is absolutely right - it cant regenerate this method, as long it does not exists.
```js
this.notifyNestedSubs = this.subscription.notifyNestedSubs.bind(this.subscription)
```
React-redux will refresh binding on hot-reload, but this is not transparent for RHL.
For now you have to live with this error. Next, we have a few options how to fix it:
 - override .bind for all class members, to detect when one call bind, and be able to understand result (hack)
 - introduce `fixtures` - a way to force-patch one classes and ignore others (leads to hell).

**Comment by zerobias** [comment]
Dirty hack

```js
function ignoreHotLoaderErrors(...methods) {
  const originalErrorLog = console.error.bind(console)
  function error(...args) {
    if (args.length >= 5 && methods.indexOf(args[4]) > -1) {
      return undefined
    }
    originalErrorLog(...args)
  }
  console.error = error
}
ignoreHotLoaderErrors('notifyNestedSubs')
```

**Comment by noinkling** [comment]
@theKashey 
I can confirm the first error was fixed in beta 4.

It looks like you already know what's going on with `notifyNestedSubs`, but just in case it's still helpful, debugging reveals that:

- Both `prevAttr` and `nextAttr` are bound, and have a `.length` of `0`
- There is no `notifyNestedSubs` on `ProxyComponent.prototype`

> introduce `fixtures` - a way to force-patch one classes and ignore others (leads to hell).

Since the `connect` HoC from react-redux is extremely common, would this be so bad?

Out of interest I tried modifying react-redux to use an arrow function instead, the error disappeared but the component still didn't re-render.

**Comment by fwh1990** [comment]
I found that rhl still worked even if I remove babel-plugin: `react-hot-loader/babel` from .babelrc,

Just occur error message but make no difference
```js

React-stand-in: Updated class  Connect(HeaderComponent) contains native or bound function  notifyNestedSubs ƒ notifyNestedSubs() {
    this.listeners.notify();
  } . Unable to reproduce, use arrow functions instead.

```

**Comment by theKashey** [comment]
@zerobias - yep, something like `ignoreHotLoaderErrors(someRules)`. Or reuse existing `warnings` flag just to hide all warnings.

@noinkling - good to know. Thank you.

@fwh1990 - RHL may work without babel plugin, but will be unable to hot-swap any arrow or bound function.

**Comment by fwh1990** [comment]
conflict with decorator by antd

https://ant.design/components/form/#Form.create(options)

```js
@hot(module)
@Form.create()
@connect()
export class SomeComponent extends Component {}

```

```js

Uncaught TypeError: this.getForm is not a function
    at ProxyComponent.render (createBaseForm.js:501)
    at ProxyComponent.render (createClassProxy.js:70)
    at render (hotReplacementRender.js:98)
    at hotReplacementRender (hotReplacementRender.js:134)
    at next (hotReplacementRender.js:151)
    at flow.forEach (hotReplacementRender.js:175)
    at Array.forEach (<anonymous>)

```

antd version:  3.0.3
rhl    version:  beta4

Any idea?

**Comment by neoziro** [comment]
@fwh1990 decorators are still an issue in v4 (#279, #727)

@theKashey I think we should add `warnings: false` by default and switch it to a `debug` option.

**Comment by theKashey** [comment]
@neoziro - or not throw anything from node_modules, ie from the non-controlled code base.

**Comment by neoziro** [comment]
@theKashey you know warnings are annoying, the only use-case I see is to debug it so I prefer to not active them by default.

**Comment by theKashey** [comment]
So we need `dev` mode for a dev tool.

**Comment by neoziro** [comment]
Fixed in [v4.0.0-beta.7](https://github.com/gaearon/react-hot-loader/releases/tag/v4.0.0-beta.7)
